app/src/main/java/com/example/drawingactivity/DrawActivity.kt

```diff
@@ -89,88 +90,117 @@ class DrawActivity : AppCompatActivity() {
-                                            index, newLayerName,
-                                            DialogOperation.EDIT
+                                        onLayerListUpdate?.invoke(
+                                            index, newLayerName, DialogOperation.EDIT
                                         )
                                     } else {
-                                        bottomSheetLambda?.invoke(
-                                            it.editLayerIndex, newLayerName, DialogOperation.EDIT
+                                        onLayerListUpdate?.invoke(
+                                            it.listIndex, newLayerName, DialogOperation.EDIT
                                         )
                                     }
                                 } else viewModel.dialogCompleted()
                             },
                         )
                     })
-                if (it.isDialogOpen && !isEditDialogOpen) {
-                    isEditDialogOpen = true
-                    dialogUtils.showDialog(
-                        context = this,
-                        layerList = it.layerList,
-                        onEditText = { layerName -> viewModel.setCurrentText(layerName) },
-                        previousInput = it.currentNewLayerName,
-                        onSubmit = { newLayerName ->
-                            if (newLayerName != null) {
-                                viewModel.editLayer(newLayerName, it.editLayerIndex)
-                                bottomSheetLambda?.invoke(
-                                    it.editLayerIndex, newLayerName,
-                                    DialogOperation.EDIT
-                                )
-                            } else viewModel.dialogCompleted()
-                        },
-                    )
-                } else if (!it.isDialogOpen) {
-                    isEditDialogOpen = false
-                }
-            } else if (!it.isLayerSheetOpen) {
-                isLayerListViewSheetOpen = false
-            }
+            } else isLayerListViewSheetOpen = false
+            if (it.isDeleteDialogOpen && !isDeleteDialogOpen) {
+                isDeleteDialogOpen = true
+                dialogUtils.confirmationDialog(
+                    context = this@DrawActivity,
+                    title = this@DrawActivity.getString(R.string.deleteLayerDialogTitle),
+                    message = this@DrawActivity.getString(
+                        R.string.deleteLayerDialogBody, it.layerList[it.listIndex]
+                    ),
+                    onConfirm = {
+                        viewModel.removeFromList(it.listIndex)
+                        onLayerListUpdate?.invoke(
+                            viewModel.getCurrentIndex(), null, DialogOperation.DELETE
+                        )
+                        viewModel.dialogCompleted()
```

Comment:
Again this and the line a few lines up are bleeding logic. Why not
`viewModel.deleteLayerConfirmed(it.listIndex)`
and in that function it calls logic to say that the dialog is completed. Try to route things in a way so that the viewmodel does most of the work in a very hands off way

Reply:
Rerouted my logic for this